Push refactor

benartuso opened the pull request.

gschoeni commented:
@benartuso I am not seeing the background tasks kicked off locally 🤔 is there any other config I have to do for redis?

gschoeni commented on Dockerfile:
```diff
@@ -49,5 +49,6 @@ FROM debian:bullseye-slim AS runtime
 WORKDIR /oxen-server
 COPY --from=builder /usr/src/oxen-server/target/release/oxen-server /usr/local/bin
 ENV SYNC_DIR=/var/oxen/data
+ENV REDIS_URL=localhost:6379
```
make sure to add the redis://

benartuso commented:
@gschoeni updated:
- factored out common redis connection logic 
- added `redis://` prefix in code and dockerfile 
- added more informative message to the connection error on server boot if can't connect to redis 
- added retries on failed `poll_until_synced` - I assumed here that the redis cache won't get cleared (aka only robust to a restart of oxen-server) but can pull the retries up a level to re-run the missing entries check if you think that's necessary!

benartuso commented:
@gschoeni Updated! 
- Made a generic `TaskQueue` enum w/ redis and in-memory variants. Makes the actual usage in the relevant parts of the codebase like a bajillion times cleaner so good call 
- Just used a `VecDeque` for the in-memory one but saw that you had been using an async `deadqueue` in some other places, so lmk if that's a better fit here 
- yay

gschoeni commented on src/server/src/main.rs:
```diff
@@ -38,6 +48,36 @@ async fn main() -> std::io::Result<()> {
+                    task.run();
+                }
+                None => {
+                    log::debug!("No queue items found, sleeping");
```
Can we comment this out for now?

gschoeni commented on src/lib/src/core/index/pusher.rs:
```diff
@@ -136,179 +224,219 @@ async fn remote_is_ahead_of_local(
+                log::debug!("Got n unsynced commits {:?}", sync_status.num_unsynced);
+                bar.set_position(commits_to_sync - sync_status.num_unsynced as u64);
+                if sync_status.num_unsynced == 0 {
+                    bar.finish();
```
Might want an extra newline here so that the ✅ pops with more whitespace around it.

lewazo commented:
Formatted, this string looks like

```
Error: Oxen CLI out of date. Pushing to OxenHub requires version >= {min_oxen_version:?}, found version {local_oxen_version:?}.

Visit https://docs.oxen.ai/getting-started/intro for update instructions.

On Mac:

  brew update
  brew upgrade oxen

On Ubuntu:

  For x86-64:
    wget https://github.com/Oxen-AI/Oxen/releases/latest/download/oxen-linux-x86_64.deb
    sudo dpkg -i oxen-linux-x86_64.deb

  For ARM64:
    wget https://github.com/Oxen-AI/Oxen/releases/latest/download/oxen-linux-arm64.deb
    sudo dpkg -i oxen-linux-arm64.deb
```
